test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/DefaultHtmlGeneratorTest.cs

```diff
@@ -190,6 +191,103 @@ public void GenerateTextArea_WithNullExpression_WithNameAttribute_DoesNotThrow()
             Assert.Equal(expected, attribute.Value);
         }
 
+        [Theory]
+        [MemberData(nameof(GetMaxLengthAttributeRenderingTestsData))]
+        public void Generate_RendersMaxLengthAttribute_AsRequested(
+            string expression,
+            int expectedValue,
+            Action<Dictionary<string, object>> customizeHtmlAttributes,
+            bool maxLengthShouldBeRendered,
+            Func<IHtmlGenerator, ViewContext, ModelExplorer, Dictionary<string, object>, string, TagBuilder> getTagBuilder)
+        {
+            // Arrange
+            var metadataProvider = new TestModelMetadataProvider();
+            var htmlGenerator = GetGenerator(metadataProvider);
+            var viewContext = GetViewContext<ModelWithMaxLengthMetadata>(model: null, metadataProvider: metadataProvider, new HtmlHelperOptions { MaxLengthAttributeRenderingEnabled = true });
+            var modelMetadata = metadataProvider.GetMetadataForProperty(typeof(ModelWithMaxLengthMetadata), expression);
+            var modelExplorer = new ModelExplorer(metadataProvider, modelMetadata, null);
+            var htmlAttributes = new Dictionary<string, object>
+            {
+                { "name", "testElement" },
+            };
+
+            customizeHtmlAttributes?.Invoke(htmlAttributes);
+
+            // Act
+            var tagBuilder = getTagBuilder(htmlGenerator, viewContext, modelExplorer, htmlAttributes, expression);
+
+            // Assert
+            if (maxLengthShouldBeRendered)
+            {
+                var attribute = Assert.Single(tagBuilder.Attributes, a => a.Key == "maxlength");
+                Assert.Equal(expectedValue, Int32.Parse(attribute.Value));
+            }
+            else
+            {
+                Assert.DoesNotContain(tagBuilder.Attributes, a => a.Key == "maxlength");
+            }
+        }
+
+        public static TheoryData<string, int, Action<Dictionary<string, object>>, bool, Func<IHtmlGenerator, ViewContext, ModelExplorer, Dictionary<string, object>, string, TagBuilder>> GetMaxLengthAttributeRenderingTestsData
```

Comment:
Move property above test where it's used or to the top of the class. Also please add a comment describing the values so it's not necessary to jump back and forth over the data to see what the values mean.